File: 019-remove-nth-from-end/solution.py
#雙指針
from typing import Optional
class ListNode:
    def __init__(self, val, prev, next):
        self.val = val
        self.prev = prev
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def AddAtTail(self, newNode):
        if self.head is None:
            print(f"Add {newNode.val} at head")
            self.head = newNode
            self.tail = newNode
            return
        newNode.prev = self.tail
        self.tail.next = newNode
        self.tail = newNode
        print(f"Add {newNode.val} at tail")

    def printList(self, head):
        if head is None:
            return
        cur = head
        nums = []
        nums.append(cur.val)
        while cur.next is not None:
            cur = cur.next
            nums.append(cur.val)
        print(nums)

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head is None:
            return None
        dummy: Optional[ListNode] = ListNode(0, None, head)
        first = dummy
        second = dummy

        #first先移動到second + n的位置
        for i in range(0, n):
            first = first.next

        print(f"first = {first.val} " + f"second = {second.val}")

        while first.next is not None:
            first = first.next
            second = second.next
            if first is None:
                print("first is None " +  f"second = {second.val}")
                break
            print(f"first = {first.val} " + f"second = {second.val}")

        #second 會在 n 前一位
        second.next = second.next.next

        return dummy.next
    """
    ↓         ↓
    d--->1--->2--->3--->4--->5
         ↓         ↓
    d--->1--->2--->3--->4--->5
                   ↓         ↓
    d--->1--->2--->3--->4--->5
                        ↓         ↓
    d--->1--->2--->3--->4--->5--->None
    """

if __name__ == "__main__":
    nums = [1, 2, 3, 4, 5]
    n = 4
    mainList = LinkedList()
    for i in nums:
        newNode = ListNode(i, None, None)
        mainList.AddAtTail(newNode)
    mainList.head = Solution().removeNthFromEnd(mainList.head, n)
    mainList.printList(mainList.head)

File: 019-remove-nth-from-end/solution.go
// 自己寫的
package main

type ListNode struct {
	val  int
	Next *ListNode
}
type LinkedList struct {
	head *ListNode
	tail *ListNode
	size int
}

func (ll *LinkedList) Append(val int) {
	Node := &ListNode{val, nil}
	if ll.head == nil {
		ll.head = Node
		ll.tail = Node
	} else {
		ll.tail.Next = Node
		ll.tail = Node
	}
	ll.size++
}
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head.Next == nil {
		return nil
	}
	dummy := &ListNode{0, head} //& ??
	l, r := dummy, dummy
	for i := 0; i < n && r.Next != nil; i++ {
		r = r.Next
	}
	for r.Next != nil {
		r = r.Next
		l = l.Next
	}
	l.Next = l.Next.Next
	return dummy.Next //注意不是return head
}
func main() {
	ll := LinkedList{nil, nil, 0}
	head := []int{1, 2, 3, 4, 5}
	for _, v := range head {
		ll.Append(v)
	}
	removeNthFromEnd(ll.head, 2)
}

/*
思路與python版本一致
019-remove-nth-from-end/solution.py

注意第39行, 一開始寫的是return head, 自己沒發現錯了
*/
